velox/type/Type.h

```diff
@@ -658,34 +634,75 @@ class DecimalType : public ScalarType<KIND> {
     return scale_;
   }
 
+  const char* kindName() const override {
+    return "DECIMAL";
+  }
+
   std::string toString() const override {
     return fmt::format("DECIMAL({},{})", precision_, scale_);
   }
 
   folly::dynamic serialize() const override {
     auto obj = ScalarType<KIND>::serialize();
+    obj["type"] = kindName();
     obj["precision"] = precision_;
     obj["scale"] = scale_;
     return obj;
   }
 
+ protected:
+  DecimalType(const uint8_t precision, const uint8_t scale)
+      : precision_(precision), scale_(scale) {
+    VELOX_CHECK_LE(
+        scale,
+        precision,
+        "Scale of decimal type must not exceed its precision");
+    VELOX_CHECK_LE(
+        precision,
+        kMaxPrecision,
+        "Precision of decimal type must not exceed {}",
+        kMaxPrecision);
+    VELOX_CHECK_GE(
+        precision,
+        kMinPrecision,
+        "Precision of decimal type must be at least {}",
+        kMinPrecision);
+  }
+
  private:
   const uint8_t precision_;
   const uint8_t scale_;
 };
 
-using ShortDecimalType = DecimalType<TypeKind::SHORT_DECIMAL>;
-using LongDecimalType = DecimalType<TypeKind::LONG_DECIMAL>;
+class ShortDecimalType : public DecimalType<TypeKind::BIGINT> {
+ public:
+  ShortDecimalType(int precision, int scale)
+      : DecimalType<TypeKind::BIGINT>(precision, scale) {}
+};
+
+class LongDecimalType : public DecimalType<TypeKind::HUGEINT> {
+ public:
+  LongDecimalType(int precision, int scale)
+      : DecimalType<TypeKind::HUGEINT>(precision, scale) {}
+};
 
-inline bool isDecimalKind(TypeKind typeKind) {
-  return (
-      typeKind == TypeKind::SHORT_DECIMAL ||
-      typeKind == TypeKind::LONG_DECIMAL);
+TypePtr DECIMAL(uint8_t precision, uint8_t scale);
+
+FOLLY_ALWAYS_INLINE bool isShortDecimalType(const Type& type) {
+  return dynamic_cast<const ShortDecimalType*>(&type) != nullptr;
 }
 
-bool isDecimalName(const std::string& typeName);
+FOLLY_ALWAYS_INLINE bool isLongDecimalType(const Type& type) {
+  return dynamic_cast<const LongDecimalType*>(&type) != nullptr;
+}
 
-bool isDecimalTypeSignature(const std::string& arg);
+FOLLY_ALWAYS_INLINE bool isDecimalType(const Type& type) {
```

Comment:
Do you think checking (kindName() == DECIMAL && TypeKind == HUGEINT) for isDecimalType would be cleaner versus checking for dynamic casts?

@mbasmanova Are we going to have PhysicalTypeKind and LogicalTypeKind. For instance, LONG_DECIMAL is Logical type and the underlying physical type is HUGEINT.

Reply:
@karteekmurthys I'm working through converting INTERVAL_DAY_TIME to logical type in #4406.

My current thinking is that TypeKind enum describes physical type and an instance of Type class describes logical type. I do not plan to have enums for logical types as these should be extensible.

I expect LongDecimalType to be logical type backed by TypeKind::HUGEINT physical type and ShortDecimalType to be logical type backed by TypeKind::BIGINT physical type.